### ISSUE TITLE
Bug with Seeded Searchers

### ISSUE THREAD

**Issue by wistuba** [awslabs/syne-tune#430]
Opened on behalf of @timyber:

When we are setting a fixed seed, random sampling always have the same behaviors. and It would be running out of search space if we run a large number of training jobs.
This can be reproduced by testing large budget (e.g. max_training_jobs: 100, batch_size: 1) and setting the seed to a fixed value.

**Comment by mseeger** [comment]
Can you please make it clearer what this issue really is here? Thanks

**Comment by mseeger** [comment]
I am happy to work on this, but need a clear motivation of what is wrong (I do not understand the text above), and ideally an example that reproduces a supposed error.

**Comment by timyber** [comment]
Context: Sagemaker AMT plans to launch a new feature called "Reproducible HPO", which will give customers the ability to set the random seed. When using the same seed, we should generate the same hyper parameters.
But we got a problem with the fixed seed:
1. we got  no enough candidates error if we request over 50 candidates
2. We found that the SyneTune can only give us 50 candidates (50 batches).
3. the problem is in SyneTune: https://code.amazon.com/packages/SyneTuneAMT/blobs/heads/SyneTuneAMT-0.11.3-full/--/third-party-src/syne_tune/optimizer/schedulers/searchers/gp_fifo_searcher.py#L63
4. The pattern SyneTune generates candidates with fixed seed is like this:
         The 1st batch: 1
	The 2nd batch: 1, 2
	The 3rd batch: 1, 2, 3
	The 4th batch: 1, 2, 3, 4
when using a fixed seed, random sampling always behaves same. So we need to increase the Random Retries.

**Comment by mseeger** [comment]
OK, so if your search space only contains 50 distinct configs, this is all you are going to get, because our schedulers never suggest a config more than once. Is that the problem?

**Comment by mseeger** [comment]
Why would it help to increase the number of random retries? They will just as well reject you picking the same config more than once. This is a hard constraint.

**Comment by hfurkanbozkurt** [comment]
We are good to close the issue as we will address this in AMT side.

**Comment by mseeger** [comment]
Just a comment about this task. If you intend to make GP-based HPO fully reproducible, than good luck with that. It involves gradient based optimization, using codes we do not fully control. These things are just not deterministic.

It would be really good to talk to scientists who wrote all of this, before coming up with tasks. We can help derisk these things, and make sure your goals are the right goals.

**Comment by hfurkanbozkurt** [comment]
Let's just cover random search first. This is required for the first step to make it reproducible.

**Comment by timyber** [comment]
> Maybe your problem is this? Say my search space has 50 distinct configs, and you ask for 50, and the current code does not sample all of them? If that is the problem, this can be fixed, in a better way than just increasing the random retries.

Yes, when we need more than 50 distinct configs. now it will throw an empty config.  Increasing the number 50 would solve the issue for now, and I think the 50 is just too small in this case.

> Just a comment about this task. If you intend to make GP-based HPO fully reproducible, than good luck with that. It involves gradient based optimization, using codes we do not fully control. These things are just not deterministic.
>
> It would be really good to talk to scientists who wrote all of this, before coming up with tasks. We can help derisk these things, and make sure your goals are the right goals.
We've already done 100+ experiments, and the result is (when setting a fixed seed):
100% reproducibility for Random and Hyperband strategies.
Improved reproducibility (50% but it depends) for Bayesian strategy.

**Comment by mseeger** [comment]
This is a pretty poor idea. You can make things reproducible by setting a master seed and generate seeds for each call from that one, using a normal PRNG.

Instead, your solution would mean that random search is always, for the same search space, going through exactly the same sequence of configs. If I ran random search over 100 datasets with xgboost search space, I get the same configs every time. That would be VERY concerning, please do not do that

**Comment by timyber** [comment]
> This is a pretty poor idea. You can make things reproducible by setting a master seed and generate seeds for each call from that one, using a normal PRNG.
> 
> Instead, your solution would mean that random search is always, for the same search space, going through exactly the same sequence of configs. If I ran random search over 100 datasets with xgboost search space, I get the same configs every time. That would be VERY concerning, please do not do that

We only need larger number of Exclusion List when it's for Reproducible HPO. for others, we change the seed every time.
If we want 100% reproducibility, then we will need another Exclusion List to maintain the state. and It will be more complicated.
